be/queue.c

```diff
@@ -0,0 +1,443 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2015-2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+
+/**
+ * @addtogroup be
+ *
+ * @{
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_BE
+#include "lib/trace.h"
+
+#include "be/queue.h"
+
+#include "lib/errno.h"          /* ENOMEM */
+#include "lib/memory.h"         /* M0_ALLOC_ARR */
+#include "lib/buf.h"            /* m0_buf_memcpy */
+#include "lib/misc.h"           /* ergo */
+
+#include "be/op.h"              /* m0_be_op_active */
+
+
+struct be_queue_item {
+	uint64_t         bqi_magic;
+	/** bqq_tl, m0_be_queue::bq_q */
+	struct m0_tlink  bqi_link;
+	char            *bqi_data[];
+};
+
+#define BE_QUEUE_ITEM2BUF(bq, bqi) \
+	M0_BUF_INIT((bq)->bq_cfg.bqc_item_length, &(bqi)->bqi_data)
+
+struct be_queue_wait_op {
+	struct m0_be_op      *bbo_op;
```

Comment:
There is very little in this queue that depends to BE. It would be better to move queue in lib/ (use channels for completion signals) and leave in be/ a small wrapper binding queue m0_be_op.

Reply:
I had the same thoughts.
There is a /lib/queue.h in the existing code.
If this is an enhancement, or this can replace the existing queue library, we should make it a general queue library.

Reply:
@nikitadanilov: this is true, but we don't have another synchronisation primitive as convenient as `m0_be_op` (`M0_BE_OP_SYNC()` family of macros, `m0_be_op_tick_ret()` etc.). This is why `m0_be_op` had been chosen.

@huanghua78: `lib/queue.h` is about completely different thing, so there is nothing to merge.